Roleplay plugin refactor.

GTG3000 opened the pull request:
Changed bio and racerole to use RSArgumentParser.
Rewrote storage to use a dataclass handling basic checks.

GTG3000 committed:
Roleplay plugin refactor.\nChanged bio and racerole to use RSArgumentParser.\nRewrote storage to use a dataclass handling basic checks.

GTG3000 committed:
Minor refactoring.

rubellyte commented on roleplay.py:
```diff
@@ -8,22 +8,118 @@
+        self.bios = self.config_manager.get_plugin_config_file\
+            ("bios.json",
+             json_save_args={'default': lambda o: asdict(o), 'indent': 2, 'ensure_ascii': False},
+             json_load_args={'object_hook': lambda o: self.Bio(**o) if {*o} == {*self.Bio.fields} else o})
```
I'd prefer to do something more like `self.Bio(**o) if o.get("__bio__")` rather than comparatively expensive set casts. I understand that this was done for compatibility with the old bios file, but that should be, ideally, handled with a conversion check on activation...

GTG3000 replied:
This also makes the JSON a bit cleaner for manual editing and simplifies some stuff like dumping and object creation since that'd be a useless "__bio__" field to have in the object. 

Also this is still not entirely compatible with old bios because old bios could skip fields and this one expects all the fields. Should probably go back and write a proper conversion logic.

rubellyte replied:
I suppose that's valid. I was just thinking, it'd be nice to have a relatively standardized way of (de)serializing objects in JSON, something like a `__classhint__` field. I dunno, something to think about later I suppose.

rubellyte commented on roleplay.py:
```diff
@@ -75,32 +171,53 @@ class Roleplay(BasePlugin):
-                        self.plugin_config[gid]["race_roles"].remove(t_role.id)
-                await respond(msg, "**AFFIRMATIVE. Roles removed from race list.**")
+            args['add'] = [r for r in [find_role(msg.guild, r) for r in args['add']] if r]
+            args['remove'] = [r for r in [find_role(msg.guild, r) for r in args['remove']]if r]
```
Tiny style violation: space needed between `]` and `if`

rubellyte commented on roleplay.py:
````diff
@@ -75,32 +171,53 @@ class Roleplay(BasePlugin):
+                if l_rem:
+                    _s += "Removed roles:\n- " + "\n- ".join(l_rem) + "\n"
+                _s += "```"
+                await respond(msg, _s)
````
More verbose variable strings would be preferred here; something like `result_str`, `role_add_list`, `role_remove_list` for example.

rubellyte commented on roleplay.py:
```diff
@@ -150,281 +267,181 @@ class Roleplay(BasePlugin):
-        elif t_field not in self.fields:
-            raise CommandSyntaxError(f"Invalid field. Available fields: {', '.join(self.fields[1:])}.")
-        self.bios.save()
+            raise CommandSyntaxError("No bio id given")
```
CommandSyntaxError messages should end with periods

rubellyte commented on roleplay.py:
```diff
@@ -150,281 +267,181 @@ class Roleplay(BasePlugin):
-             syntax="(attach the file to the message, no arguments required)",
+             doc="Parses a json file or a json codeblock to update/create character bios.\n"
+                 "See output of ",
+             syntax="(attach file to the message, or put json contents into a code block following the command)",
```
JSON is an acronym and should be capitalized

rubellyte reviewed (changes requested):
Looks good for the most part, but some minor changes are in order

GTG3000 committed:
Quick update pass on levelling plugin.
Removed unneeded default config options.
Fixed error with bot trying to parse xp in channels it can't read.

GTG3000 committed:
Implemented requested changes.

Signed-off-by: GTG3000 <gtg3013@gmail.com>

GTG3000 committed:
Merge remote-tracking branch 'upstream/master'

GTG3000 committed:
I said, implemented the requested changes.

Signed-off-by: GTG3000 <gtg3013@gmail.com>
